File: problems-originals/0701-0800/0720_longest-word-in-dictionary/solution_trie_dfs.ts
interface TrieNode {
    children: Map<string, TrieNode>;
    end: boolean;
}

function longestWord(words: string[]): string {
    // The trie stores every word once; a node's `end` marks where a word
    // stops. Walking only through `end` nodes keeps every spelled prefix
    // a word, so each path the walk takes is a buildable word.
    const root: TrieNode = { children: new Map(), end: false };
    for (const word of words) {
        let node = root;
        for (const character of word) {
            let child = node.children.get(character);
            if (child === undefined) {
                child = { children: new Map(), end: false };
                node.children.set(character, child);
            }
            node = child;
        }
        node.end = true;
    }
    let best = "";
    const walk = (node: TrieNode, path: string): void => {
        // Strictly longer wins; among equal lengths the smaller word
        // wins — compared explicitly, never via child order.
        if (path.length > best.length || (path.length === best.length && path < best)) {
            best = path;
        }
        node.children.forEach((child, character) => {
            if (child.end) {
                walk(child, path + character);
            }
        });
    };
    walk(root, "");
    // Nothing buildable at all: the statement's empty-string answer.
    return best;
}

File: problems-adapt/0701-0800/0720_longest-buildable-word/solution_trie_dfs.rs
use std::collections::HashMap;

struct TrieNode {
    children: HashMap<u8, TrieNode>,
    end: bool,
}

impl Solution {
    pub fn longest_buildable_word(words: Vec<String>) -> String {
        // The trie stores every word once; a node's `end` marks where a word
        // stops. Walking only through `end` nodes keeps every spelled prefix
        // a word, so each path the walk takes is a buildable word. The
        // alphabet is lowercase ASCII, so raw bytes key the children.
        let mut root = TrieNode {
            children: HashMap::new(),
            end: false,
        };
        for word in &words {
            let mut node = &mut root;
            for byte in word.bytes() {
                node = node.children.entry(byte).or_insert(TrieNode {
                    children: HashMap::new(),
                    end: false,
                });
            }
            node.end = true;
        }
        let mut best = String::new();
        walk(&root, &mut String::new(), &mut best);
        // Nothing buildable at all: the statement's empty-string answer.
        best
    }
}

fn walk(node: &TrieNode, path: &mut String, best: &mut String) {
    // Strictly longer wins; among equal lengths the smaller word
    // wins — compared explicitly, never via child order.
    if path.len() > best.len() || (path.len() == best.len() && path.as_str() < best.as_str()) {
        *best = path.clone();
    }
    for (byte, child) in &node.children {
        if child.end {
            path.push(*byte as char);
            walk(child, path, best);
            path.pop();
        }
    }
}
